tests/magics/tlambda.nim

```diff
@@ -0,0 +1,351 @@
+  const tfun3a = alias2 tfun3
+  count = 0
+  tfun3a()
+  doAssert count == 1
+  block:
+    # a local variable `count` doesn't hijack original one
+    var count = 12
+    tfun3a()
+    doAssert count == 12
+  doAssert count == 2
+
+  ## with a proc (with side effects)
+  proc tfun4() = count.inc
+  const tfun4a = alias2 tfun4
+  tfun4()
+  doAssert count == 3
+
+  doAssert type(tfun4a) is proc # usage indistinguishable from aliased symbol
+
+  ## with an un-mapped variable count2
+  template tfun5() =
+    count2.inc
+
+  block:
+    var count2 = 12
+    const tfun5a = alias2 tfun5
+    const tfun5b = alias2 tfun5a # alias of alias
+    tfun5a()
+    doAssert count2 == 13
+    tfun5b()
+    doAssert count2 == 14
+
+  block:
+    const T = alias2 int # works with types
+    const T2 = alias2 system.int # works with fully qualified
+    const system2 = alias2 system # works with fully qualified
+    const T3 = alias2 system2.int
+    doAssert T is int
+    doAssert T2 is int
+    doAssert T3 is int
+
+  block:
+    doAssert mbar(5, alias2 tfun) == ("mbar", 5, tfun(5))
+    doAssert mbar2(5, alias2 tfun) == ("mbar", 5, tfun(5))
+
+proc testAliasReturn() =
+  proc genAlias(x: static int): aliassym =
+    when x == 1: result = alias2 tfun
+    elif x == 2: alias2 tfunDifferent
+    elif x == 3: result = alias2 tfun2
+    else:
+      static: doAssert false
+
+  const y1 = alias2 genAlias(1)
+  const y2 = alias2 genAlias(2)
+  doAssert y1(4) == tfun(4)
+  doAssert y2(4) == tfunDifferent(4)
+  doAssert y1(4) != y2(4) # sanity check
+  const z = alias2 tfun2
+  doAssert z() == tfun2()
+  const y3 = alias2 genAlias(3)
+  doAssert y3() == tfun2()
+
+proc testAliasTuple() =
+  block:
+    const funs = (alias2 tfun, alias2 tfun2)
+    doAssert funs[0](3) == tfun(3)
+    doAssert funs[1]() == tfun2()
+  block:
+    const funs = (a: alias2 tfun, b: alias2 tfun2)
+    doAssert funs[0](3) == tfun(3)
+    doAssert funs.a(3) == tfun(3)
+    const funa = alias2 funs.a
+    const funb = alias2 funs.b
+    doAssert funa(3) == tfun(3)
+    doAssert funb() == tfun2()
+
+proc testAliasFields() =
+  # this could be improved, by supporting:   `type B = object: fun1: aliassym fun2: aliassym`
+  type A[T1, T2] = object
+    fun1: T1
+    fun2: T2
+  proc initA[T1, T2](fun1: T1, fun2: T2): A[T1, T2] =
+    A[T1, T2](fun1: alias2 fun1, fun2: alias2 fun2)
+  const a = initA(a~>a*10, a~>a*20)
+  doAssert a.fun1(3) == 30
+  doAssert a.fun2(3) == 60
+
+proc testLambdaIt() =
+  const fun = lambdaIt (it, it)
+  doAssert fun(3) == (3, 3)
+  doAssert fun("foo") == ("foo", "foo")
+
+  icounter = 0
+  doAssert fun(identity("foo")) == ("foo", "foo")
+  doAssert icounter == 1  # side effect safe
+  block:
+    template fun2(it): untyped = (it, it) # not side effect safe
+    icounter = 0
+    doAssert fun2(identity("foo")) == ("foo", "foo")
+    doAssert icounter == 2 # side effect executed each time argument is used
+
+proc testArrow() =
+  block: # 1 arg
+    const fun = a ~> (a, a)
```

Comment:
Since this creates a template, would it suffer from the double evaluation problem of templates parameter? In other words, it I pass an expression that produces side effects (e.g. printing something to screen), it will be evaluated twice during the construction of the tuple.

Reply:
there are tests for that, it's side effect safe:
```nim
block:
  const fun = a ~> (a, a)
  var count = 0
  proc identity(b: int): int = (count.inc; b)
  doAssert fun(identity(3)) == (3,3)
  doAssert count == 1
```
(`~>` is a pure library implementation on top of `alias`, see std/lambdas)